libbeat/metric/system/diskio/diskstat_linux.go

```diff
@@ -52,6 +55,28 @@ func (stat *IOStat) OpenSampling() error {
 	return stat.curCPU.Get()
 }
 
+// a few of the diskio counters are actually 32-bit on the kernel side, which means they can roll over fairly easily.
+// Here we try to reconstruct the values by calculating the pre-rollover delta from unt32 max, then adding.
+// If you want to get technical, this could be a tad unsafe, as we don't actually have any way of knowing if the word size changes in a future kernel, and we've rolled over at UINT64_MAX
+
+// See https://docs.kernel.org/admin-guide/iostats.html and https://github.com/torvalds/linux/blob/master/block/genhd.c
+func returnOrFix(current, prev uint64) uint64 {
+	var maxUint32 uint64 = math.MaxUint32 //4_294_967_295 Max value in uint32/unsigned int
+
+	if current >= prev {
+		return current - prev
+	}
+	// we're at a uint64 if we hit this
+	if prev > maxUint32 {
```

Comment:
If the underlying value is 32 bits, this will never happen?

If it's 64 bits, wouldn't you want to do the same math with `math.MaxUint64`?

The only time you will need to do any math that depends on the actual integer width is when rollover occurs. Aren't you guaranteed that prev > maxUint32 in that case, and you can switch the max for the calculation to use maxUint64 instead?

Reply:
Yah, that felt like a paranoid edge case that I should in theory cover, but I wasn't sure if it made sense to actually bother. I mean, if we actually want to "fix" 64-bit rollover, that function should be on all fields, not just a few.

Reply:
Fair enough, in that case let's put 32 in the function name so that it's obvious that we are only trying to fix this for 32 bit counters.

`returnOrFix32BitRollover` or something like that.

Reply:
The safest way to address this would be to change the signature to `func(curr, prev uint32) uint64`, bit widening within the function. This would also have the advantage of statically removing the possibility of this case.

Reply:
++ That is much better than just changing the name of the function, or hoping this doesn't happen.

Reply:
It does have consequences that concern me; if the kernel changes type for this in the future, the call will still work but with silently corrupted results. What you want to cover that case is non-wrap-around integer arithmetic which would require the full width to be known at call time. This would either be (with ugliness) passing both the 32-bit truncation and the 64-bit original, or with a type conversion helper that signals the unexpected high bits somehow — the calling functions return an error, so this is possible, but with four calls it starts to get unwieldy.

Just thought I should bring these up to avoid sending you down the wrong rabbit hole.